core/primitives/src/transaction.rs

```diff
@@ -211,6 +212,61 @@ impl BorshDeserialize for Transaction {
     }
 }
 
+/// Using the new type pattern to construct a type of signed transaction that is
+/// guaranteed to have various checks performed on it.  In particular, ensure
+/// that the signature is verified and the max transaction size checks have been
+/// conducted.
+#[derive(Debug)]
+pub struct ValidatedTransaction(SignedTransaction);
+
+impl ValidatedTransaction {
+    #[allow(clippy::result_large_err)]
+    pub fn new(
+        config: &RuntimeConfig,
+        signed_tx: SignedTransaction,
+    ) -> Result<Self, (InvalidTxError, SignedTransaction)> {
+        // Don't allow V1 currently. This will be changed when the new protocol version is introduced.
+        if matches!(signed_tx.transaction, Transaction::V1(_)) {
+            return Err((InvalidTxError::InvalidTransactionVersion, signed_tx));
+        }
+        let tx_size = signed_tx.get_size();
+        let max_tx_size = config.wasm_config.limit_config.max_transaction_size;
+        if tx_size > max_tx_size {
+            return Err((
+                InvalidTxError::TransactionSizeExceeded { size: tx_size, limit: max_tx_size },
+                signed_tx,
+            ));
+        }
+
+        if signed_tx
+            .signature
+            .verify(signed_tx.get_hash().as_ref(), signed_tx.transaction.public_key())
+        {
+            Ok(Self(signed_tx))
+        } else {
+            Err((InvalidTxError::InvalidSignature, signed_tx))
+        }
+    }
+
+    pub fn to_signed_tx(&self) -> &SignedTransaction {
+        &self.0
+    }
+
+    pub fn into_signed_tx(self) -> SignedTransaction {
+        self.0
+    }
+
+    pub fn to_tx(&self) -> &Transaction {
+        &self.0.transaction
+    }
+
+    /// This method should only be used for test purposes.  This is because
+    /// kv_runtime is not designed to do proper signature verification.
+    pub fn new_for_test(signed_tx: SignedTransaction) -> Self {
```

Comment:
Any way this could be `#[cfg(test)]` gated or something like that?

Style: usually the constructors/static methods come at the beginning of the `impl` block(s). While we don't actively peruse rustdoc documentation for our crates, the order of the generated documentation is directly influenced by the order in the source code and people used to the convention generally expect to find the constructors within the group of the first couple methods documented.

Reply:
already tried `cfg(test)`.  Unfortunately, the way kv runtime is setup, it cannot see this constructor.

I will move the constructor to the top of the impl block.